### PR TITLE
negated types

### LINKED ISSUE #139: not operator (anything but)
tl;dr:

```
(param: Any) => not String
// shorthand
(param: Any) => !String

// union case
(param: Any) => not(String | Array)
// short hand
(param: Any) => !(String | Array)
```

Sometimes it's way easier and convenient to define the type by discarding one or several.
There's also the possibility to define it as a substraction—e.g. `Any - String`—but that would not be very conventional.

### TypeScript approaches

- substraction: https://github.com/Microsoft/TypeScript/issues/4183
- exclude: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#predefined-conditional-types
- negation: https://github.com/Microsoft/TypeScript/issues/18280

### PR DESCRIPTION
**Pull request by Mouvedia**
close #139

### CODE REVIEW

**Comment by Mouvedia** [comment]
Iv added minor changes from #136.

**Comment by Mouvedia** [comment]
Iv added minor changes from #136.
You will have to squash the commits when merging.

**Comment by Mouvedia** [comment]
Iv added minor changes from #136.
@ericelliott You will have to squash the commits.

**Comment by Mouvedia** [comment]
If you don't like my example Iv got another one:

`beforeSend([Function...] | Function) => false | !false`

Albeit it wouldn't introduce `&`.

File: README.md [hunk 1]
```diff
@@ -294,7 +295,29 @@ For **∅ or more** and **1 or more** element(s) of the same type you can use th
 Union types are denoted with the pipe symbol, `|`:
 
 ```js
-(userInput: String|Number) => String|Number
+(userInput: String | Number) => String | Number
+```
+
+### Negated Types
+
+It is sometime easier and more informative to delimit a type by defining what it's not. The negation operator let's you exclude by substracting from `Any`.
```
Comment (ericelliott) [inline README.md]: /let's/lets/

File: README.md [hunk 2]
```diff
@@ -294,7 +295,29 @@ For **∅ or more** and **1 or more** element(s) of the same type you can use th
...
+  => Boolean | Number | String | Object | Array | null,
+  throws SyntaxError
+
+// is less concise than
+
+JSON::parse(String, reviver: Function)
+  => !Function & !Void & !Symbol,
+  throws SyntaxError
+
+// which is equivalent to
```
Comment (ericelliott) [inline README.md:316]: This example is not equivalent. `&` indicates a single value which satisfies the interface for all of the specified types, where as the above example would reject any value that satisfies the `Symbol` type, regardless of whether or not it also satisfied `Void` and `Function`.

The example below can't be satisfied by any type, because nothing can be simultaneously satisfy `Void` and `Symbol` and `Function` (if a Void return was evaluated, it would be `undefined`).

**Review by ericelliott — CHANGES_REQUESTED** [review]
See comments.

File: README.md (hunk 2 above)
Comment (ericelliott) [reply README.md:316]: This would hold if you swapped out the `&` for `|` in the second example.

**Review by ericelliott — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (Mouvedia) [reply README.md]: _is not Function_ *and* _is not Void_ *and* _is not Symbol_ (it's an intersection > it must be `&`)
equals
_is Boolean_ *or* _is Number_ *or* _is String_ *or* _is Object_ *or* _is Array_ *or* _is null_

With 2 caveats:

1. *is Object* may have different interpretations
2. it's highly unlikely but a new builtin might be introduced (e.g. Symbol was)

File: README.md [hunk 3]
```diff
@@ -294,7 +295,29 @@ For **∅ or more** and **1 or more** element(s) of the same type you can use th
...
+
+JSON::parse(String, reviver: Function)
+  => !(Function & Void & Symbol),
+  throws SyntaxError
 ```
```
Comment (Mouvedia) [inline README.md:321]: > This example

Are you talking about [this one](https://github.com/ericelliott/rtype/pull/140/files#diff-04c6e90faac2675aa89e2176d2eec7d8R318)?
You are right then.

File: README.md [hunk 4]
```diff
@@ -294,7 +295,29 @@ For **∅ or more** and **1 or more** element(s) of the same type you can use th
...
+
+```js
+JSON::parse(String, reviver: Function)
+  => Boolean | Number | String | Object | Array | null,
+  throws SyntaxError
+
+// is less concise than
+
+JSON::parse(String, reviver: Function)
+  => !Function & !Void & !Symbol,
```
Comment (Mouvedia) [inline README.md:313]: 2 caveats:

1. `Object` may have different interpretations
2. it's highly unlikely but a new builtin might be introduced (e.g. `Symbol` was)

File: README.md (hunk 2 above)
Comment (ericelliott) [reply README.md:316]: That is correct for the first example: `!Function & !Void & !Symbol` - an intersection of the negated types.

However, the second example: `!(Function & Void & Symbol)` does not mean the same thing.

That one means is not (Function and Void and Symbol) -- where `Function & Void & Symbol` are an intersection of the non-negated types, which means to be rejected by the second example, a value would have to satisfy the interface of all three, which is impossible.

In other words, the first example rejects `Function`, `Void` and `Symbol`.

The second example rejects nothing.

**Comment by Mouvedia** [comment]
@ericelliott your comments have been addressed.

**Comment by Mouvedia** [comment]
@ericelliott **reminder** always squash the commits when merging. I didn't think you wouldn't so I didn't do it.

File: README.md [hunk 5]
```diff
@@ -294,7 +295,29 @@ For **∅ or more** and **1 or more** element(s) of the same type you can use th
...
+(userInput: String | Number) => String | Number
+```
+
+### Negated Types
+
+It is sometime easier and more informative to delimit a type by defining what it's not. The negation operator lets you exclude by substracting from `Any`.
+
+```js
+JSON::parse(String, reviver: Function)
+  => Boolean | Number | String | Object | Array | null,
```
Comment (Mouvedia) [inline README.md:307]: @ericelliott there's a problem with this example: `Array` and `Object` are not precise enough. These objects and arrays must **only** contain JSON values; i.e. it's recursive.

The example needs to be updated.
